handout/forms/form-builder/form-builder_custom_validation.md

```diff
@@ -0,0 +1,102 @@
+# FormBuilder Custom Validation
+
+As useful as the built-in validators are, it is very useful to be able to include your own.
+Angular 2 allows you to do just that, with minimal effort.
+
+Let's assume we are using the same Login Form, but now we also want to test that our password has an exclamation mark somewhere in it.
+
+_app/login-form.component.ts_
+```ts
+function hasExclamationMark (input: FormControl) {
+  const hasExclamation = input.value.indexOf('!') >= 0;
+
+  return hasExclamation ? null : { needsExclamation: true };
+}
+
+
+// ...
+this.password = new FormControl('', [
+  Validators.required,
+  hasExclamationMark
+]);
+```
+
+I've created a simple function which takes the `FormControl` instance and returns null if everything is fine.
+If the test fails, it returns an object with an arbitrarily named property.
+The property name is what will be used for the `.hasError()` test.
+
+_app/login-form.component.ts_
+```html
+<!-- ... -->
+<div [hidden]="!password.hasError('needsExclamation')">
+  Your password must have an exclamation mark!
+</div>
+<!-- ... -->
+```
+
+### Predefined Parameters
+Having a custom validator check for exclamation marks might be helpful, but what if you need to check for some other form of punctuation?
+It might be overkill to write nearly the same thing over and over again.
+
+Consider the earlier example `Validators.minLength(5)`.
+How did they get away with allowing you to pass in an argument to control the length?
+Simple, really.
+It's not a trick of Angular, or TypeScript.
+It's simple JavaScript closures.
+
+```javascript
+function minLength (minimum) {
+  return function (input) {
+    return input.value >= minimum ? null : { minLength: true };
+  };
+}
+```
+
+I have a function which takes a "minimum" parameter, and returns another function. The function defined and returned from the inside becomes my validator. My closure reference allows me to remember the value of the minimum, when the validator is eventually called.
+
+
+Let's apply that thinking back to a `PunctuationValidator`.
+
+_app/login-form.component.ts_
+```ts
+function hasPunctuation (punctuation: string, errorType: string) {
+  return function (input: FormControl) {
+    return input.value.indexOf(punctuation) >= 0 ?
+        null :
+        { [errorType]: true };
+  };
+}
+
+// ...
+
+this.password = new FormControl('', [
+  Validators.required,
+  hasPunctuation('&', 'ampersandRequired')
+]);
+```
+
+_app/login-form.component.html_
+```html
+<!-- ... -->
+<div [hidden]="!password.hasError('ampersandRequired')">
+  You must have an &amp; in your password.
+</div>
+<!-- ... -->
+```
+
+### Validating Multiple Inputs
+Keep in mind what was mentioned earlier: inputs have access to their parent context via `.root`.
+Therefore, complex validation can happen by drilling through the form, via root.
+
+```ts
+function duplicatePassword (input: FormControl) {
```

Comment:
had issues getting this to work, as input.root.controls has the potential to be undefined at a given point.

Working version:

```js
import {Component} from '@angular/core';
import {
  REACTIVE_FORM_DIRECTIVES,
  FormBuilder,
  FormControl,
  Validators
} from '@angular/forms';

function hasPunctuation (punctuation: string, errorType: string) {
  return function (input: FormControl) {
    return input.value.indexOf(punctuation) >= 0 ?
        null :
        { [errorType]: true };
  };
}

function duplicatePassword (input: FormControl) {
 
 if(!input.root.controls) {
   return null;
 }
 
  const exactMatch = input.root.controls.password.value === input.value;
  
  return exactMatch ? null : { mismatchedPassword: true };
}


@Component({
  selector: 'login-form',
  templateUrl: 'app/login-form.component.html',
  directives: [REACTIVE_FORM_DIRECTIVES]
})
export class LoginForm {
  loginForm: FormGroup;
  username: FormControl;
  password: FormControl;
  duplicatePassword: FormControl;
  
  constructor (builder: FormBuilder) {
    this.username = new FormControl('', [
      Validators.required,
      Validators.minLength(5)
    ]);
    
    this.password = new FormControl('', [Validators.required, hasPunctuation('&','ampersandRequired')]);
    this.duplicatePassword = new FormControl('', [
      Validators.required,
      duplicatePassword
    ]);
    
    this.loginForm = builder.group({
      username: this.username,
      password: this.password,
      duplicatePassword : this.duplicatePassword
    });
  }
  login () {
    console.log(this.loginForm.value);
    // Attempt Logging in...
  }
}
```